File: test-rigs/python-test-rig/taskmanager/repository.py
"""Task repository interface and in-memory implementation."""

from abc import ABC, abstractmethod
from threading import Lock
from typing import Optional

from .model import NotFoundError, Priority, Task


class TaskRepository(ABC):
    """Abstract repository for task persistence."""

    @abstractmethod
    def add(self, task: Task) -> Task:
        ...

    @abstractmethod
    def get(self, task_id: int) -> Optional[Task]:
        ...

    @abstractmethod
    def get_all(self) -> list[Task]:
        ...

    @abstractmethod
    def update(self, task: Task) -> Task:
        ...

    @abstractmethod
    def remove(self, task_id: int) -> bool:
        ...

    @abstractmethod
    def find_by_tag(self, tag: str) -> list[Task]:
        ...

    @abstractmethod
    def get_by_priority(self, priority: Priority) -> list[Task]:
        ...


class InMemoryTaskRepository(TaskRepository):
    """Thread-safe in-memory task repository."""

    def __init__(self) -> None:
        self._tasks: dict[int, Task] = {}
        self._next_id = 1
        self._lock = Lock()

    def add(self, task: Task) -> Task:
        with self._lock:
            stored = Task(
                id=self._next_id,
                title=task.title,
                description=task.description,
                priority=task.priority,
                completed=task.completed,
                tags=list(task.tags),
                created_at=task.created_at,
                completed_at=task.completed_at,
            )
            self._tasks[self._next_id] = stored
            self._next_id += 1
            return stored

    def get(self, task_id: int) -> Optional[Task]:
        with self._lock:
            return self._tasks.get(task_id)

    def get_all(self) -> list[Task]:
        with self._lock:
            return list(self._tasks.values())

    def update(self, task: Task) -> Task:
        with self._lock:
            if task.id not in self._tasks:
                raise NotFoundError(f"Task {task.id} not found")
            self._tasks[task.id] = task
            return task

    def remove(self, task_id: int) -> bool:
        with self._lock:
            if task_id in self._tasks:
                del self._tasks[task_id]
                return True
            return False

    def find_by_tag(self, tag: str) -> list[Task]:
        with self._lock:
            return [t for t in self._tasks.values() if tag in t.tags]

    def get_by_priority(self, priority: Priority) -> list[Task]:
        with self._lock:
            return [t for t in self._tasks.values() if t.priority == priority]

File: test-rigs/go-test-rig/repository/repository.go
package repository

import (
	"sort"
	"strings"
	"sync"

	"github.com/jkjamies/omnivore/go-test-rig/model"
)

// TaskRepository defines the interface for task persistence.
type TaskRepository interface {
	GetAll() []*model.Task
	GetByID(id uint64) (*model.Task, error)
	Add(task *model.Task) error
	Update(task *model.Task) error
	Remove(id uint64) (*model.Task, error)
	FindByTag(tag string) []*model.Task
	Count() int
}

// InMemoryTaskRepository is an in-memory implementation for testing.
type InMemoryTaskRepository struct {
	mu     sync.RWMutex
	tasks  map[uint64]*model.Task
	nextID uint64
}

// NewInMemoryTaskRepository creates a new in-memory repository.
func NewInMemoryTaskRepository() *InMemoryTaskRepository {
	return &InMemoryTaskRepository{
		tasks:  make(map[uint64]*model.Task),
		nextID: 1,
	}
}

// NextID returns the next available ID.
func (r *InMemoryTaskRepository) NextID() uint64 {
	r.mu.Lock()
	defer r.mu.Unlock()
	id := r.nextID
	r.nextID++
	return id
}

// GetAll returns all tasks sorted by ID.
func (r *InMemoryTaskRepository) GetAll() []*model.Task {
	r.mu.RLock()
	defer r.mu.RUnlock()
	tasks := make([]*model.Task, 0, len(r.tasks))
	for _, t := range r.tasks {
		clone := *t
		tasks = append(tasks, &clone)
	}
	sort.Slice(tasks, func(i, j int) bool {
		return tasks[i].ID < tasks[j].ID
	})
	return tasks
}

// GetByID returns a task by its ID.
func (r *InMemoryTaskRepository) GetByID(id uint64) (*model.Task, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	task, ok := r.tasks[id]
	if !ok {
		return nil, model.ErrNotFound(id)
	}
	clone := *task
	return &clone, nil
}

// Add stores a new task, rejecting duplicate titles.
func (r *InMemoryTaskRepository) Add(task *model.Task) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, existing := range r.tasks {
		if strings.EqualFold(existing.Title, task.Title) {
			return model.ErrDuplicateTitle(task.Title)
		}
	}
	clone := *task
	r.tasks[task.ID] = &clone
	return nil
}

// Update replaces an existing task.
func (r *InMemoryTaskRepository) Update(task *model.Task) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	if _, ok := r.tasks[task.ID]; !ok {
		return model.ErrNotFound(task.ID)
	}
	clone := *task
	r.tasks[task.ID] = &clone
	return nil
}

// Remove deletes a task and returns it.
func (r *InMemoryTaskRepository) Remove(id uint64) (*model.Task, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	task, ok := r.tasks[id]
	if !ok {
		return nil, model.ErrNotFound(id)
	}
	delete(r.tasks, id)
	return task, nil
}

// FindByTag returns all tasks that have the given tag (case-insensitive).
func (r *InMemoryTaskRepository) FindByTag(tag string) []*model.Task {
	r.mu.RLock()
	defer r.mu.RUnlock()
	tagLower := strings.ToLower(tag)
	var result []*model.Task
	for _, t := range r.tasks {
		for _, tt := range t.Tags {
			if strings.ToLower(tt) == tagLower {
				clone := *t
				result = append(result, &clone)
				break
			}
		}
	}
	return result
}

// Count returns the number of tasks.
func (r *InMemoryTaskRepository) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.tasks)
}
